### PR TITLE
Replace docs with auto-generated OpenAPI docs

### LINKED ISSUE #2964: [BUG] Inconsistency with trailing slashes with API URLs in documentation sample code, 404 when trailing slash is present
**Describe the bug**
The URLs provided in the sample code in the API Documentation [sometimes contain trailing slashes](https://uclapi.com/docs/#dashboard/api/analytics/services). When running the sample code, we would get 404 because of the trailing slash, whereas it would work if the trailing slash is removed from the URL.

**To Reproduce**
Steps to reproduce the behavior (using one of the endpoints):
1. request https://uclapi.com/dashboard/api/analytics/methods/
2. Get 404

**Expected behavior**
A response from the API (not a 404)

**Additional context**
Django usually uses trailing slashes with all their URLs, however there is significant controversy regarding whether an URI with a trailing slash is the same as another URI _without_ a trailing slash ([discussion 1](https://stackoverflow.com/questions/61547014/restful-uri-trailing-slash-or-no-trailing-slash), [discussion 2](https://github.com/open-policy-agent/opa/issues/972#issuecomment-427937639)). 

By default, Django [has a setting that redirects URLs](https://docs.djangoproject.com/en/dev/ref/settings/?from=olddocs#append-slash) that don't match any urlconf to the same URL plus a trailing slash. Adding a trailing slash to the URL patterns in the respective `urls.py` would likely solve the issue, however the documentation states that:
> Note that the redirect **may** cause any data submitted in a POST request to be lost.

and I'm not sure how that affects the endpoints that do use POST. @greenfrogs

### LINKED ISSUE #3054: OpenAPI Schema
**Is your feature request related to a problem? Please describe.**
Currently we have to edit the frontend manually to create docs, which offers no checking for us and is quite tedious.

**Describe the solution you'd like**
OpenAPI schema in this repo!

**Describe alternatives you've considered (optional)**
The status quo.

### LINKED ISSUE #3055: [Feature Request] Auto-generate documentation with OpenAPI
**Is your feature request related to a problem? Please describe.**
Writing documentation for APIs is annoying...

**Describe the solution you'd like**
Statically generate the documentation based on the schema (yay no javascript ;) ).

**Describe alternatives you've considered (optional)**
The status quo, it sucks.

### PR DESCRIPTION
**Pull request by shu8**
## What does this PR do?
Replaces the manually-written documentation with automatically Swagger-rendered docs from our OpenAPI specification at https://github.com/uclapi/uclapi-openapi/blob/master/uclapi.json.

Unless I've missed anything, the new Swagger docs should have everything that the old docs had **except for** the example snippets for each endpoint.

The old example snippets used to be mostly the same for each endpoint, so if we wanted to keep this, maybe we could just add a single example for each type at the top?

A cURL example is automatically generated by Swagger when you try the API via the UI though.

image

image

Closes #3055. Closes #3054.

## ✅ Pull Request checklist

- [ ] Is this code complete?
- [ ] Are tests done/modified?
- [ ] Are docs written for this PR? (if applicable)

## 🚨 Is this a breaking change for API clients?
No

## :squirrel: Deploy notes
N/A

## Anything else
Fixes #2964

### CODE REVIEW

**Comment by hueyy** [comment]
Some of the code examples were a bit special, like the OAuth stuff, which I think it might be useful to keep. Is there any way we can include some of those code snippets, maybe in the description or something?

**Comment by Feverfew** [comment]
> Some of the code examples were a bit special, like the OAuth stuff, which I think it might be useful to keep. Is there any way we can include some of those code snippets, maybe in the description of each endpoint or something?

I agree the OAuth workflow should be something we can keep.

A couple of questions I'd like clarified:
1) One thing is that the SDKs should always correspond to what is currently running on the server(s), so the deploy scripts for staging and production should be where the new SDK is pushed from. This sounds like it requires two SDKs, as the code on staging and production is different. Is it possible to take this into account?
2) Related to the above, I'd prefer if the definition was in this repo (and sounds like it will make it easier to take into account the first point). Firstly, this will mean that its easier to enforce that changes in the API are also reflected in changes to the OpenAPI definition at the same time (i.e. in the same PR). Secondly, it sounds like that it will make it easier to meet the requirements of the point I made above.

Otherwise, I like the fact that we've got rid of this many lines of code!

Could we also also have documentation in the `backend/source` directory to explain the recommended workflow in regards to editing the schema?

**Comment by shu8** [comment]
I've updated the spec to add code samples for the OAuth endpoints, and added one example at the top for how to call the roombookings api, which hopefully people will find easy to tweak for the other endpoints: https://github.com/uclapi/uclapi-openapi/commit/c79149fa4148d5c47e61d764f1b59c5db6941cc8

---

I think adding the spec file to this repo makes sense for both the reasons you mentioned, but just a quick (possibly stupid) question: how/when do we currently promote staging to production?

Having separate SDKs should also work, and it might be enough to just create separate branches in the SDK repos and publishing to npm, etc. with different version numbers depending on which one it is.

**Comment by Feverfew** [comment]
> I think adding the spec file to this repo makes sense for both the reasons you mentioned, but just a quick (possibly stupid) question: how/when do we currently promote staging to production?

Whenever Zak feels like it, manually. Should be automated with Github Actions, after integration tests on staging are run.

> Having separate SDKs should also work, and it might be enough to just create separate branches in the SDK repos and publishing to npm, etc. with different version numbers depending on which one it is.
Yeah the workflow I'm thinking of is.
1. Master is updated.
2. Deployment to staging. Once deployed, job is triggered to create new staging SDK(s) based on OpenAPI spec in the repo and push them to NPM automatically (potentially with some smoke tests?).
3. Same for production, but obviously a different SDK. Whether it's a different branch or repo, I'm not too worried about.

**Comment by Feverfew** [comment]
As discussed in the meeting once the OpenAPI spec is added to this repo and the documentation is built from it, then I'm happy to review and merge this. We can deal with SDK stuff in a separate PR

File: frontend/src/pages/Documentation.jsx
Comment (Feverfew) [inline frontend/src/pages/Documentation.jsx:13]: Can this be built statically? The reason I ask is, firstly for performance reasons, but more importantly is for consistency. At what point is `https://cdn.jsdelivr.net/gh/uclapi/uclapi-openapi/uclapi.json` updated? It seems like either the documentation should be generated as part of a build step or be part of the statics?

**Review by Feverfew — CHANGES_REQUESTED** [review]
One question that I'd like to hash out before approval

File: frontend/src/pages/Documentation.jsx
Comment (shu8) [reply frontend/src/pages/Documentation.jsx:13]: Yeah I agree statically building is best -- I should have really marked this as WIP after our meeting sorry! 

I haven't had much time to look into the best way to build it statically and join it with the webpack build yet, but I'll try looking into it ASAP - openapi-generator has a [dynamic-html docs generation option](https://openapi-generator.tech/docs/generators/dynamic-html) which looks promising for us!

Comment (shu8) [reply frontend/src/pages/Documentation.jsx]: It's part of the build process now!

**Comment by shu8** [comment]
@Feverfew I've:

- added some sphinx docs for the openapi documentation process,
- moved the openapi docs rendering to be a build-time step so it is always in sync with the current deployment, and
- updated the spec to add descriptions to all the fields (which we had in the existing docs but I forgot to copy over when I did the openapi spec originally!)

if it looks good then I will move onto the SDK deployment work and archive the old `uclapi-openapi` repo (https://github.com/uclapi/uclapi-openapi)!

**Review by Feverfew — DISMISSED** [review]
LGTM let's get this in staging.

**Comment by greenfrogs** [comment]
We need to wait for staging to be fixed first, and #3052 to merge. Possibly also adding the new docs from #3052 (otherwise we are going to have some really annoying merge conflicts on the other request).

**Comment by shu8** [comment]
> Possibly also adding the new docs from #3052 (otherwise we are going to have some really annoying merge conflicts on the other request).

I'll add those into this PR now, so once workspaces is merged we can get this in quicker!

**Comment by shu8** [comment]
I've added the docs from #3052 into this PR but the review got dismissed with the new push! Please could either of you have another quick look @greenfrogs @Feverfew!

**Comment by shu8** [comment]
sorry for the delay but all merge conflicts and tests should be fixed now for the final re-review! @greenfrogs @Feverfew

**Comment by zipy124** [comment]
@uclapi/ucl-api-team are we ready to get this into master now?

**Comment by shu8** [comment]
@zipy124 I think it is -- the last review got dismissed because of all the merge conflicts I fixed and the new workspaces docs but nothing else has changed! 

I'll resolve this package-lock.json conflict now too!

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:11]: nit: do we want to match this version to some sort of release of the API?

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:576]: nit: ac.uk

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:610]: ooc: any reason for `int64` instead of `number`?

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:729]: nit: can these numbers be outside of quotes like above (or do floats cause problems)?

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:762]: super nit: remove .0

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:777]: nit: number in quotes

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:947]: nit: number in quotes

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:1111]: ooc: do we return time zone here? Our API seems to be really inconsistent 😅

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:1092]: This might be a mistake in the API but does this return a string rather than a number? (same for `survey_device_id`)

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:1144]: nit: integer instead of number (same with `averages`)

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:1187]: nit: int

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:1340]: nit: same as above string for number value

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:2137]: nit: is this a string? is this a good example?

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:2795]: nit: I think this is a mistake in the API, I believe we format date/time as datetime

File: uclapi.openapi.json
Comment (greenfrogs) [inline uclapi.openapi.json:4251]: nit: can we call this `Production server`?

**Review by greenfrogs — APPROVED** [review]
@shu8 sorry for the very late review. This is an excellent piece of work and will make it so much easier to support and use the API in the future.

I have some really small comments but they are mostly only nits so I am happy to approve for merging when you feel it is best. Lots of my comments are likely problems with the underlying API or weird inconsistencies which have risen as a result of difficulty in seeing datatypes/phrasing for the entire API. Feel free to not address these and I can raise issues to address them in the future.

**Comment by shu8** [comment]
Thanks @greenfrogs! Yeah those all seem like things we should fix!

I should have time to go over these (and go through all the endpoints fully once more to see if there is any other weirdness!) over the next week but it might be good to get this into master/staging and I can submit another PR for all the fixes?

**Comment by zipy124** [comment]
> Thanks @greenfrogs! Yeah those all seem like things we should fix!
> 
> I should have time to go over these (and go through all the endpoints fully once more to see if there is any other weirdness!) over the next week but it might be good to get this into master/staging and I can submit another PR for all the fixes?

I'm good with this decision!

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:1111]: Yeah it look like we return exactly what comes from occupeye

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:947]: Turns out we actually return lat/long as a string! that seems intentional, maybe for storage or precisions concerns?

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:777]: this is also actually a string, updated!

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:11]: Yes! We can do that along with #3379!

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:1092]: Yeah looks like we're returning a string in the API, again using the value straight from Occupeye

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:1144]: I think it might be better to be more specific and stick with `integer` here?

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:1187]: same with this! what do you think?

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:1340]: I think this is a string because it's a HTML SVG so all the attributes would be strings

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:2795]: updated error message to be more clear: `start_datetime/end_datetime isn't formatted as suggested in the docs`

File: uclapi.openapi.json
Comment (shu8) [reply uclapi.openapi.json:1144]: Looking through the code we have a lot of fields with types that I don't really get, e.g., `weeknumber` is a `FloatField` for `/roombookings`, and `BigIntegerField` for `/timetable`.

we could probably update a lot of our types -- what do you think? if we did #3060 this could go along with it.

**Pull request merged**
